File: C++/3.无重复字符的最长子串.cpp

// 3. 无重复字符的最长子串
//
// 给定一个字符串，请你找出其中不含有重复字符的最长子串的长度。

#include <algorithm>
#include <map>
#include <string>

class Solution {
public:
    // Offer 48
    // 双指针, 滑动窗口, 同时使用map存储字符出现过的下标
    int lengthOfLongestSubstring(std::string s) {
        std::map<char, int> hash;  // 用于存每个字符和它的下标
        int max_len = 0;           // 最大长度

        // 左右指针遍历字符串
        for (int l = 0, r = 0; r < s.size(); r++) {
            if (hash.find(s[r]) != hash.end()) {
                // 若该字符已经存在, 将左边界向前移动, 不能直接移动到index+1, 原因: abba
                l = std::max(l, hash[s[r]] + 1);
            }

            // 若map中存在则更新最新下标, 若不存在则将kv插入
            hash[s[r]] = r;

            // 计算最大长度
            max_len = std::max(max_len, r - l + 1);
        }

        return max_len;
    }
};

File: leetcode/04字符串/3. 无重复字符的最长子串.go
package main

// 3. 无重复字符的最长子串

// 给定一个字符串 s ，请你找出其中不含有重复字符的 最长子串 的长度。

// lengthOfLongestSubstring .
// 双指针, 滑动窗口, 同时使用map存储字符出现过的下标
// 滑动窗口的的字符所构成的子串就是每一个无重复字符的子串
func lengthOfLongestSubstring(s string) int {
	sbyte := []byte(s)
	hash := make(map[byte]int) // 存储字符和它最近一次出现的下标
	maxLen := 0                // 无重复字符的最长子串长度

	// 双指针遍历字符串形成滑动窗口，同时根据窗口长度计算无重复字符的最长子串长度
	for l, r := 0, 0; r < len(sbyte); r++ {
		// 若当前字符已经出现过，将左边界向前移动或保持不变，不能直接移动到index+1，原因: abba
		// 当前字符上次出现的位置在窗口内也可能不在窗口内，若在窗口内那么需要向前移动左指针到index+1，若没在窗口内那就不需要移动左指针
		if index, ok := hash[sbyte[r]]; ok {
			l = max(l, index+1)
		}

		// 记录或更新当前字符最近一次出现的下标
		hash[sbyte[r]] = r

		// 更新无重复字符的最长子串长度
		maxLen = max(maxLen, r-l+1)
	}

	return maxLen
}

// max .
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
